01-resources-v2/proposal.md

```diff
@@ -137,40 +202,99 @@ type PutRequest struct {
+  remain in the database for archival purposes, but it will no longer be a
+  candidate for any builds.
+
+  Required fields for this event:
 
-`put` allows new spaces to be generated dynamically (based on params
-and/or the bits in its working directory) and propagated to the rest of the
-pipeline.
+  * `space`: The space the version is in.
+  * `version`: The version object.
 
-Note that a `put` may only affect one space at a time, otherwise it becomes
+Because the space is included on each event, `put` allows a new space to be
+generated dynamically (based on params and/or the bits in its working
+directory) and propagated to the rest of the pipeline. However it must take
+care to only affect one space at a time. Without this restriction it becomes
 difficult to express things like "`get` after `put`" to fetch the version that
 was created. If multiple spaces are returned, it's unclear which space the
 `get` would fetch from.
 
-Versions returned with `created` type will be recorded as outputs of the build.
-A `check` will then be performed to fill in the metadata and determine the
-ordering of the versions. Once the ordering is learned, the latest version will
-be fetched by the implicit `get`.
+#### the `get` after the `put`
 
-Versions returned with `deleted` type will be marked as deleted. They will
-remain in the database for archival purposes, but will no longer be input
-candidates for any builds, and can no longer be fetched. The implicit `get`
-after `deleted` puts should not happen so this will result in changes to that
-behavior.
+With v1 resources, every `put` implied a `get` of the version that was created.
+With v2 we will change that, so that the `get` is opt-in. This has been a
+long-time ask, and one objective reason to make it opt-in is that Concourse
+can't know ahead of time that there will even be anything to `get` - for
+example, the `put` could emit only `deleted` events.
 
-Anything printed to `stdout` and `stderr` will propagate to the build logs.
+So, to `get` the latest version that was produced by the `put`, you would
+configure something like:
+
+```yaml
+- put: my-resource
+  get: my-created-resource
+- task: use-my-created-resource
+```
+
+The value for the `get` field is the name of the artifact to save. When
+specified, the last version emitted will be fetched.
+
+This added flexibility enables resources to provide explicitly versioned
+'variants' of original versions without doubling up the version history. One
+use case for this is pull-requests: you may want a build to pull in one
+resource for the PR itself, another resource for the base branch of the
+upstream reap, and then `put` to produce a "combined" version of the two,
+representing the PR merged into the upstream repo:
+
+```yaml
+jobs:
+- name: run-pr
+  plan:
+  - get: concourse-pr  # pr: 123, ref: deadbeef
+    trigger: true
+  - get: concourse     # ref: abcdef
+  - put: concourse-pr
+    get: merged-pr
+    params:
+      merge_base: concourse
```

Comment:
Ref your comment in https://github.com/telia-oss/github-pr-resource/issues/5#issuecomment-440416235, I figured I'd leave my questions here! 

Given the documentation in the RFC, presumably we must have a `default_space` in the source configuration for our resources, or specify `spaces: []` for each `get`:

```yml
- name: run-pr
  plan:
  - get: concourse-pr
    spaces: [ master ] # <- Must be configured here or in the source right?
    trigger: true
  - get: concourse     
    spaces: [ master ] # <- Same
  - put: concourse-pr
    get: merged-pr
```

And based on the above/If i've understood your suggestion correctly, it seems that with Resources V2 the `github-pr-resource` should emit versions like this:


```json
{
  "action": "discovered",
  "space": "master",
  "version": { 
    "pr": 123,
    "ref": "deadbeef"
  },
  "metadata": [{...}]
}
```

This necessarily means that our job `name: run-pr` can only ever test PRs against a single branch at a time. So if someone is running e.g. GitFlow or something similar with multiple branches, they would have to add a job for each base (branch, e.g. `develop` and `master`)? 

This assumes that we **can't** do `spaces: [develop, master]` on `get`, but even if we could, we'd also need complex logic in Concourse to figure out that a PR against `develop` should use the latest version from the `develop` space of `get: concourse` when creating `merged-pr` - which I presume is not there?

Reply:
PS: The last paragraph before the "example"-section states that:
> the user must specify it explicitly in the pipeline, either by configuring one on the resource (default_space: foo) or on every get step using the resource (spaces: [foo]).

I believe `spaces: [foo]` should be `space: foo` according to the documentation for `get` further down? I'll update my comment above to reflect this.

Reply:
@itsdalmo `get` steps specify arrays of spaces (and/or patterns like `feature/*` for all `feature/`-prefixed branches). You could specify `[develop, master]` there but I'm not sure that'd be very useful. Sorry, I'm having kind of a hard time parsing your question - is the concern that the base branch for a PR varies from PR to PR and you'd need some way to coordinate that with this approach?